voucher/views.py

```diff
@@ -1,3 +1,180 @@
+        else:
+            return super().get(request, *args, **kwargs)
+
+    def form_valid(self, form):
+        voucher = get_current_voucher(self.request.user)
+        course_run = get_object_or_404(CourseRun, id=form.cleaned_data["course_run"])
+        return redirect_to_checkout(voucher, course_run)
+
+    def get_form(self, form_class=None):
+        form = super().get_form(form_class=form_class)
+        voucher = get_current_voucher(self.request.user)
+        course_matches = match_courses_to_voucher(voucher)
+        form.fields["course_run"].choices = [
+            (match.id, match.title) for match in course_matches
+        ]
+        return form
+
+    def get_context_data(self, **kwargs):
+        return {
+            **super().get_context_data(**kwargs),
+            **get_js_settings_context(self.request),
+        }
+
+
+@login_required
+def resubmit(request):
+    """
+    Prompt user to email voucher after failed voucher parsing
+    """
+    return render(request, "resubmit.html", context=get_js_settings_context(request))
+
+
+@login_required
+def redeemed(request):
+    """
+    Coupon has already been redeemed
+    """
+    return render(request, "redeemed.html", context=get_js_settings_context(request))
+
+
+@login_required
+def invalid(request):
+    """
+    There are no valid coupons for this voucher
+    """
+    return render(request, "invalid.html", context=get_js_settings_context(request))
+
+
+def redirect_to_checkout(voucher, course_run):
+    """
+    Get coupon eligibility and redirect to the checkout page
+    """
+    product = Product.objects.filter(course_run=course_run).first()
+
+    valid_coupon_versions = get_valid_coupon_versions(product, voucher.user)
+    if not valid_coupon_versions:
+        return redirect("voucher:invalid")
+    coupon = valid_coupon_versions[0].coupon
+    voucher.coupon = coupon
+    voucher.save()
+    enroll_url = f"/enroll?product_id={product.id}&code={voucher.coupon.coupon_code}"
+    return redirect(enroll_url)
+
+
+def get_current_voucher(user):
+    """Get the active voucher for a user"""
+    return user.vouchers.order_by("created_on").last()
+
+
+def match_courses_to_voucher(voucher):
+    """
+    Try to find an exact match for the course specified by a voucher
+
+    If no exact match is found, search for partial matches. Expand this function for more complex fuzzy matching
+    """
+    exact_course_matches = CourseRun.objects.filter(
+        course__readable_id__exact=voucher.course_id_input,
+        course__title__iexact=voucher.course_title_input,
+        start_date__date=voucher.course_start_date_input,
+    )
+    if exact_course_matches.count() > 1:
+        # Handle multiple exact matches
+        return CourseRun.objects.none()
+    elif exact_course_matches.count():
+        return exact_course_matches
+    else:
+        return CourseRun.objects.filter(
+            Q(course__readable_id__exact=voucher.course_id_input)
```

Comment:
If I end up here because an exact match is not found, the voucher coupon may not be valid for some of the course runs returned here.

![Screen Shot 2019-05-24 at 9 48 03 AM](https://user-images.githubusercontent.com/187676/58335904-a38cc900-7e10-11e9-97d4-84c6bc0910fd.png)
If I pick one such class from the dropdown, I get directed to `/invalid` with the message `Sorry, there are no valid coupons available at this time.`, so it is handled gracefully.  But Peter thinks it would be better to filter out any course runs not eligible for the product first, so the user doesn't waste time selecting an invalid course run.  Something like this (there may be a more efficient query but this is what I came up with off the top of my head):

```
CourseRun.objects.filter(
            Q(course__readable_id__exact=voucher.course_id_input)
            | Q(course__title__iexact=voucher.course_title_input)
            | Q(start_date__date=voucher.course_start_date_input)
        ).filter(id__in=CouponEligibility.objects.filter(coupon__id=voucher.coupon_id).values_list("product__object_id", flat=True))
```

Reply:
Sure. Then I can remove the invalid page entirely

Reply:
Oops, sorry about that.  Your suggestion sounds good, it might be query intensive if there's lots of potential matches but that seems unlikely.